contracts/03-tamagotchi-nft/src/lib.rs

```diff
@@ -1,19 +1,146 @@
 #![no_std]
 
 #[allow(unused_imports)]
-use gstd::prelude::*;
+use gstd::{debug, exec, msg, prelude::*};
+use tamagotchi_nft_io::{Tamagotchi,TmgEvent, TmgAction};
+
+const HUNGER_PER_BLOCK: u64 = 1;
+const BOREDOM_PER_BLOCK: u64 = 2;
+const ENERGY_PER_BLOCK: u64 = 2;
+const FILL_PER_FEED: u64 = 1000;
+const FILL_PER_ENTERTAINMENT: u64 = 1000;
+const FILL_PER_SLEEP: u64 = 1000;
+
+static mut TAMAGOTCHI: Option<Tamagotchi> = None;
 
 #[no_mangle]
 extern fn init() {
     // TODO: 0️⃣ Copy the `init` function from the previous lesson and push changes to the master branch
+    let init_msg: String = msg::load().expect("Can't decode an init message");
+
+    let tamagotchi = Tamagotchi {
+        name: "Ivan".to_string(),
+        date_of_birth: exec::block_timestamp(),
+        owner: msg::source(),
+        fed: 1000,
+        fed_block: exec::block_height().into(),
+        entertained: 1000,
+        entertained_block: exec::block_height().into(),
+        slept: 1000,
+        slept_block: exec::block_height().into(),
+        approved_account: None,
+    };
+    debug!(
+        "The Tamagotchi Program was initialized with name {:?}, birth date {:?}, owner: {:?}",
+        tamagotchi.name, tamagotchi.date_of_birth, tamagotchi.owner
+    );
+    unsafe { TAMAGOTCHI = Some(tamagotchi) };
 }
 
+
 #[no_mangle]
 extern fn handle() {
     // TODO: 0️⃣ Copy the `handle` function from the previous lesson and push changes to the master branch
+    let _tamagotchi = unsafe {
+        TAMAGOTCHI
+            .as_mut()
+            .expect("The contract is not initialized")
+    };
+    
+    let name = &_tamagotchi.name;
+    let current_time = exec::block_timestamp();
+    let age = current_time - _tamagotchi.date_of_birth;
+    let action: TmgAction = msg::load().expect("Can't decode an action message");
+    
+    // 
+
+    let _event = match action {
+        TmgAction::Name => {
+            msg::reply(TmgEvent::Name(name.to_string()), 0).expect("Error in sending name");
+        }
+        TmgAction::Age =>{
+            msg::reply(TmgEvent::Age(age), 0).expect("Error in sending age");
+           
+        } 
+        TmgAction::Feed => {
```

Comment:
No check that Tamagotchi has died. 
overflow is also possible when you subtract one value from another.
You should use `saturating_sub`